File: py/ngff_zarr/parse_metadata.py
# SPDX-FileCopyrightText: Copyright (c) Fideus Labs LLC
# SPDX-License-Identifier: MIT
from typing import Optional

from .methods import Methods
from ._supported_versions import NgffVersion
from .v04.zarr_metadata import (
    Omero,
    OmeroChannel,
    OmeroWindow,
    MethodMetadata
)

def _extract_method_metadata(metadata_dict: dict) -> tuple[Optional[Methods], Optional[str], Optional[MethodMetadata]]:
    method = None
    method_type = None
    method_metadata = None
    if isinstance(metadata_dict, dict):
        if "type" in metadata_dict and metadata_dict["type"] is not None:
            method_type = metadata_dict["type"]
            # Find the corresponding Methods enum
            for method_enum in Methods:
                if method_enum.value == method_type:
                    method = method_enum
                    break

        # Extract method metadata if present
        if "metadata" in metadata_dict and metadata_dict["metadata"] is not None:
            from .v04.zarr_metadata import MethodMetadata

            metadata_dict = metadata_dict["metadata"]
            if isinstance(metadata_dict, dict):
                method_metadata = MethodMetadata(
                    description=str(metadata_dict.get("description", "")),
                    method=str(metadata_dict.get("method", "")),
                    version=str(metadata_dict.get("version", "")),
                )
    return method, method_type, method_metadata


def _parse_omero(omero_data: dict) -> Optional[Omero]:
    """Parse OMERO metadata dictionary into Omero dataclass."""
    omero = None
    if isinstance(omero_data, dict) and "channels" in omero_data:
        channels_data = omero_data["channels"]
        if isinstance(channels_data, list):
            channels = []
            for channel in channels_data:
                if not isinstance(channel, dict) or "window" not in channel:
                    continue

                window_data = channel["window"]
                if not isinstance(window_data, dict):
                    continue

                # Handle backward compatibility for OMERO window metadata
                # Some stores use min/max, others use start/end, some have both
                if "start" in window_data and "end" in window_data:
                    # New format with start/end
                    start = float(window_data["start"])  # type: ignore
                    end = float(window_data["end"])  # type: ignore
                    if "min" in window_data and "max" in window_data:
                        # Both formats present
                        min_val = float(window_data["min"])  # type: ignore
                        max_val = float(window_data["max"])  # type: ignore
                    else:
                        # Only start/end, use them as min/max
                        min_val = start
                        max_val = end
                elif "min" in window_data and "max" in window_data:
                    # Old format with min/max only
                    min_val = float(window_data["min"])  # type: ignore
                    max_val = float(window_data["max"])  # type: ignore
                    # Use min/max as start/end for backward compatibility
                    start = min_val
                    end = max_val
                else:
                    # Invalid window data, skip this channel
                    continue

                channels.append(
                    OmeroChannel(
                        color=str(channel["color"]),  # type: ignore
                        label=str(channel.get("label", None))
                        if channel.get("label") is not None
                        else None,  # type: ignore
                        window=OmeroWindow(
                            min=min_val,
                            max=max_val,
                            start=start,
                            end=end,
                        ),
                    )
                )

            if channels:
                omero = Omero(channels=channels)

    return omero

def _detect_version(root_attrs: dict) -> NgffVersion:
    """Detect NGFF version from root attributes."""
    version_str: Optional[str] = None
    if "ome" in root_attrs:
        version_str = root_attrs["ome"].get("version")
    else:
        multiscales = root_attrs.get("multiscales", [])
        if multiscales and isinstance(multiscales, list):
            version_str = multiscales[0].get("version", "0.4")
        
    if version_str is None:
        raise ValueError("Could not detect NGFF version from root attributes.")

    return NgffVersion(version_str)

File: ts/src/utils/parse_metadata.ts
// SPDX-FileCopyrightText: Copyright (c) Fideus Labs LLC
// SPDX-License-Identifier: MIT

/**
 * Utilities for parsing OME-Zarr metadata from zarr attributes.
 */

import {
  isSupportedVersion,
  NgffVersion,
} from "../types/supported_versions.ts";
import type { Methods } from "../types/methods.ts";
import type {
  MethodMetadata,
  Omero,
  OmeroChannel,
  OmeroWindow,
} from "../types/zarr_metadata.ts";

/**
 * Result from extracting method metadata
 */
export interface MethodMetadataResult {
  method: Methods | undefined;
  methodType: string | undefined;
  methodMetadata: MethodMetadata | undefined;
}

/**
 * Extract method type and convert to Methods enum.
 */
export function extractMethodMetadata(
  metadataDict: Record<string, unknown>,
): MethodMetadataResult {
  let method: Methods | undefined = undefined;
  let methodType: string | undefined = undefined;
  let methodMetadata: MethodMetadata | undefined = undefined;

  if (metadataDict && typeof metadataDict === "object") {
    if ("type" in metadataDict && metadataDict.type !== null) {
      methodType = metadataDict.type as string;
      // Find the corresponding Methods enum
      // Import dynamically to avoid circular dependency
      const methodsValues = [
        "itkwasm_gaussian",
        "itkwasm_bin_shrink",
        "itkwasm_label_image",
      ];
      if (methodsValues.includes(methodType)) {
        method = methodType as Methods;
      }
    }

    // Extract method metadata if present
    if ("metadata" in metadataDict && metadataDict.metadata !== null) {
      const metadata = metadataDict.metadata as Record<string, unknown>;
      if (metadata && typeof metadata === "object") {
        methodMetadata = {
          description: String(metadata.description ?? ""),
          method: String(metadata.method ?? ""),
          version: String(metadata.version ?? ""),
        };
      }
    }
  }

  return { method, methodType, methodMetadata };
}

/**
 * Parse OMERO metadata dictionary into Omero interface.
 */
export function parseOmero(
  omeroData: Record<string, unknown> | undefined | null,
): Omero | undefined {
  if (!omeroData || typeof omeroData !== "object") {
    return undefined;
  }

  if (!("channels" in omeroData) || !Array.isArray(omeroData.channels)) {
    return undefined;
  }

  const channels: OmeroChannel[] = [];

  for (const channel of omeroData.channels as Array<Record<string, unknown>>) {
    if (
      !channel ||
      typeof channel !== "object" ||
      !("window" in channel) ||
      !channel.window
    ) {
      continue;
    }

    const windowData = channel.window as Record<string, unknown>;
    if (typeof windowData !== "object") {
      continue;
    }

    // Handle backward compatibility for OMERO window metadata
    // Some stores use min/max, others use start/end, some have both
    let start: number;
    let end: number;
    let minVal: number;
    let maxVal: number;

    if ("start" in windowData && "end" in windowData) {
      // New format with start/end
      start = Number(windowData.start);
      end = Number(windowData.end);
      if ("min" in windowData && "max" in windowData) {
        // Both formats present
        minVal = Number(windowData.min);
        maxVal = Number(windowData.max);
      } else {
        // Only start/end, use them as min/max
        minVal = start;
        maxVal = end;
      }
    } else if ("min" in windowData && "max" in windowData) {
      // Old format with min/max only
      minVal = Number(windowData.min);
      maxVal = Number(windowData.max);
      // Use min/max as start/end for backward compatibility
      start = minVal;
      end = maxVal;
    } else {
      // Invalid window data, skip this channel
      continue;
    }

    const window: OmeroWindow = {
      min: minVal,
      max: maxVal,
      start: start,
      end: end,
    };

    const omeroChannel: OmeroChannel = {
      color: String(channel.color),
      window,
      ...(channel.label !== undefined && channel.label !== null
        ? { label: String(channel.label) }
        : {}),
      ...(typeof channel.active === "boolean"
        ? { active: channel.active }
        : {}),
    };

    channels.push(omeroChannel);
  }

  if (channels.length === 0) {
    return undefined;
  }

  return {
    channels,
    ...(typeof omeroData.version === "string"
      ? { version: omeroData.version }
      : {}),
  };
}

/**
 * Detect NGFF version from root attributes.
 */
export function detectVersion(
  rootAttrs: Record<string, unknown>,
): NgffVersion {
  let versionStr: string | undefined = undefined;

  if ("ome" in rootAttrs && rootAttrs.ome) {
    const ome = rootAttrs.ome as Record<string, unknown>;
    versionStr = ome.version as string | undefined;
  } else {
    const multiscales = rootAttrs.multiscales as unknown[];
    if (multiscales && Array.isArray(multiscales) && multiscales.length > 0) {
      const firstMultiscale = multiscales[0] as Record<string, unknown>;
      versionStr = (firstMultiscale.version as string) ?? "0.4";
    }
  }

  if (versionStr === undefined) {
    throw new Error("Could not detect NGFF version from root attributes.");
  }

  if (!isSupportedVersion(versionStr)) {
    throw new Error(`Unsupported NGFF version: ${versionStr}`);
  }

  return versionStr as NgffVersion;
}
